lib/booking.rb

```diff
@@ -0,0 +1,242 @@
+        rooms << availible[i]
+        i += 1
+      end # .times
+      return rooms
+    end # get_rooms
+
+    def make_reservation(start_date, end_date, num_rooms )
+      valid_dates(start_date, end_date)
+      if start_date < end_date
+        availible = availible_rooms(start_date, end_date)
+        if num_rooms > 20
+          raise BookingError.new("You can't reserve that many rooms because we only have 20 rooms in the hotel.")
+        elsif num_rooms > availible.length
+          raise BookingError.new("We don't have that many rooms availible for those dates. We only have #{availible.length} rooms availible for those dates. ")
+        end # if/elsif
+        dates_booked = get_date_array(start_date, end_date)
+        reservation_id = @all_reservations.length + 1
+        rooms = get_rooms(start_date, end_date, num_rooms)
+        cost = (num_rooms * dates_booked.length * 200.0)
+        @all_reservations << Hotel::Reservation.new(reservation_id, cost, rooms, dates_booked)
+      end # if/else
+    end # make_reservation
+
+
+    def check_date_for_reservations(start_date, end_date)
+      # it will have a method that checks if there is a reservation for the date(s) requested
+      # iterate though @all_reservations for each day requested and if the day is included in the Booking then it will look at the Reservation that included that date and add the room(s) in that reservation to a new array not_availible.
+      days = get_date_array(start_date, end_date)
+      date_reservations = []
+      days.each do |day|
+        @all_reservations.each do |res|
+          if res.date_range.include?(day)
+            if !(date_reservations.include?(res))
+              date_reservations << res
+            end
+          end # if
+        end # .each
+      end #.each
+      return date_reservations
+    end # check_date_for_reservations
+
+    def print_reservations(start_date, end_date)
+      # Not sure if we need this and if simply calling check_date_for_reservations is enough for the user to be able to access all the reservations for a given date(range)?
+      reservations = check_date_for_reservations(start_date, end_date)
+      nice_format_reservations = []
+      reservations.each do |res|
+        loop_array = []
+        loop_array << "Reservation ID: #{res.res_id}"
+        loop_array << "Total cost: $#{res.total_cost}"
+        list_rooms = ""
+        res.res_rooms.each do |room|
+          list_rooms << " #{room.room_number},"
+        end #.each
+        loop_array << "Rooms reserved: #{list_rooms}"
+        loop_array << "Date range: #{res.date_range[0]} - #{res.date_range[-1] + 1}"
+        nice_format_reservations << loop_array
+      end
+      return nice_format_reservations
+    end #print_reservations
+
+    def availible_rooms(start_date, end_date)
+      reservations = check_date_for_reservations(start_date, end_date)
+      blocks = check_date_for_block(start_date, end_date)
+      reserved = []
```

Comment:
I like this implementation, and appreciate the commenting. It feels like a lot of code, and it might be possible to streamline it a little, but the general idea (get a list of reservations and blocks that overlap the date range, then subtract those rooms from the list of all rooms) is easy to follow.